### ISSUE TITLE
Memory leak and crash, now and 2 years ago, tfjs-node

### ISSUE THREAD

**Issue by borodadada** [tensorflow/tfjs#8326]
The problem has not disappeared anywhere, 2 years ago and now the same thing. Visually, through the task manager in the past there was an information leak, and the process was constantly increasing in memory, now this is not the case, everything is fine, but the result is the same, after a while everything crashes. I managed to take screenshots at the moment when it all started. The screenshots show my working environment, NOT the TEST that I posted, the test itself is as close and simplified as possible, in the future I will post information on this test.
I'm not doing anything, just taking screenshots.

There are 4 identical programs running on the computer, 4 copies, one of them begins to fail. This happens when the number of epochs is measured in millions. For test you can run only one copy. On a modern processor, the procedure usually takes 4-6 hours, on an old one more than a day.

Memory leak starts, this starts happening quickly, as can be seen in the screenshot
![Snipaste_2024-07-06_07-28-22](https://github.com/tensorflow/tfjs/assets/5166757/58f80843-04ce-412e-abc0-7d5a102e6438)

Process, note 3 other processes, usually from 100 to 200 megabytes in size
![Snipaste_2024-07-06_07-28-54](https://github.com/tensorflow/tfjs/assets/5166757/a47d535c-48ec-4d42-b067-6a71f9a1f23b)

Full memory
![Snipaste_2024-07-06_07-30-42](https://github.com/tensorflow/tfjs/assets/5166757/7ee8b80c-fbad-4730-a1af-b93ebd5a1520)

After
![Snipaste_2024-07-06_07-30-58](https://github.com/tensorflow/tfjs/assets/5166757/d59ac86d-e916-4968-8ed8-ed6f71f1f7c4)

All node js have closed
![Snipaste_2024-07-06_07-31-53](https://github.com/tensorflow/tfjs/assets/5166757/ca214020-f63a-48d1-823e-b97f0d24003e)

Logs, there is nothing in them, they are empty, the editor is open
![Snipaste_2024-07-06_07-37-51](https://github.com/tensorflow/tfjs/assets/5166757/ef39638e-6076-48bc-91c6-902a2ff83d33)

For test simple code, just copy past

TEST CODE
```
const tf = require('@tensorflow/tfjs-node');

const size = 50
const units = 100

const letsgo = async function(){

    const model = tf.sequential();
    model.add( tf.layers.dense({ inputShape: [units], units, activation: 'linear', useBias: true }));
    model.add( tf.layers.dense({ units, activation: 'linear', useBias: true }));
    model.add( tf.layers.dense({ units, activation: 'linear', useBias: true }));
    model.compile({ optimizer: tf.train.adam(0.005, 0.9, 0.999), loss: tf.losses.absoluteDifference });

    let a = []
    let b = []
    for (let i = 0; i < size; i++) {
        let aa = []
        let bb = []
        for (let ii = 0; ii < units; ii++) {
            aa.push( Math.random() )
            bb.push( Math.random() )
        }
        a.push(aa)
        b.push(bb)
    }

    let xs = tf.tensor2d( a );
    let ys = tf.tensor2d( b );

    await model.fit(xs, ys, {
        epochs: 50000000,
        shuffle: false,
        verbose: 0,
        callbacks:{
            onTrainBegin: ()=>{
                console.log('start')
            },
            onTrainEnd: ()=>{
                console.log('done')
            },
            onEpochEnd: async (epoch, logs)=>{
                if( epoch % 100000 === 0 )
                    console.log(epoch, logs.loss)
            }
        }
    })
}

const loop = async function(){
    for (let i = 0; i < 1; i++) {
        await letsgo()
    }
}

loop()
```

**System information**
- Windows 11 x64
- node-v19.9.0-x64
- node-v20.15.0-x64
- "@tensorflow/tfjs": "^4.20.0",
- "@tensorflow/tfjs-node": "^4.20.0",

Okey, this is results from test code:

modern PC intel 13700, crash after 4.4 millions epochs

![Snipaste_2024-07-06_12-38-43](https://github.com/tensorflow/tfjs/assets/5166757/7439ba05-b840-4356-b52b-4a0c3ffb726c)

old PC intel 3770, crash after 4.4 millions epochs - windows 10 x64 + nodejs 20.10.0

![Snipaste_2024-07-06_16-49-30](https://github.com/tensorflow/tfjs/assets/5166757/2754b681-c26e-4e71-85b4-f7b036e524fa)

I can't do the calculations because the program always crashes, and I need many more epochs than here!!! I really hope you fix this, it's a disaster that this bug hasn't been fixed for years!

**Comment by gaikwadrahul8** [comment]
Hi, @borodadada 

I apologize for the delay in my response and thank you for bringing this issue to our attention and as far I know to avoid memory leak you'll have to use [tf.tidy](https://js.tensorflow.org/api/latest/#tidy) which executes the provided function fn and after it is executed, cleans up all intermediate tensors allocated by `fn` except those returned by `fn`. `fn` must not return a Promise (async functions not allowed). The returned result can be a complex object.

Using this method helps avoid memory leaks. In general, wrap calls to operations in [tf.tidy()](https://js.tensorflow.org/api/latest/#tidy) for automatic memory cleanup.

NOTE: Variables do not get cleaned up when inside a `tidy()`. If you want to dispose variables, please use [tf.disposeVariables()](https://js.tensorflow.org/api/latest/#disposeVariables) or call `dispose()` directly on variables please refer [tf.dispose](https://js.tensorflow.org/api/latest/#dispose).

You can also use [tf.memory](https://js.tensorflow.org/api/latest/#memory) which returns memory info at the current time in the program.

Could you please give it try after adding [tf.tidy](https://js.tensorflow.org/api/latest/#tidy) and [tf.dispose](https://js.tensorflow.org/api/latest/#dispose) in your code and see memory leak is happening or not ? 

If I have missed something here please let me know.

Thank you for your cooperation and patience.

**Comment by borodadada** [comment]
I don’t understand the principle of a memory leak, all the algorithm needs is to change the coefficients and then feedback and comparison of the result, it should work endlessly without a memory leak
What part of the code should be placed in Tidy? I posted an example

**Comment by mightyplow** [comment]
You have to dispose the xs and ys after the fit step. Otherwise tfjs creates new tensors in every loop step. They fill up the memory with each step if you don't dispose them.

**Comment by borodadada** [comment]
```
    await model.fit(xs, ys, {
        epochs: 50000000,
        shuffle: false,
        verbose: 0,
        callbacks:{
            onTrainBegin: ()=>{
                console.log('start')
		xs = undefined  // ????????????????
		ys = undefined // ????????????????
            },
            onTrainEnd: ()=>{
                console.log('done')
            },
            onEpochEnd: async (epoch, logs)=>{
                if( epoch % 100000 === 0 )
                    console.log(epoch, logs.loss)
            }
        }
    })
```

I don’t understand the logic at all, but is that what you mean?

**Comment by mightyplow** [comment]
it should be like this
`
let xs = tf.tensor2d( a );
let ys = tf.tensor2d( b );

await model.fit(xs, ys, {
        epochs: 50000000,
        shuffle: false,
        verbose: 0,
        callbacks:{
            onTrainBegin: ()=>{
                console.log('start')
            },
            onTrainEnd: ()=>{
                console.log('done')
            },
            onEpochEnd: async (epoch, logs)=>{
                if( epoch % 100000 === 0 )
                    console.log(epoch, logs.loss)
            }
        }
})

xs.dispose();
ys.dispose();
`

**Comment by borodadada** [comment]
I am writing to you about this, that it has not reached this point
The program crashes on the - model.fit

```
await model.fit(xs, ys, {
        epochs: 50000000,
        shuffle: false,
        verbose: 0,
        callbacks:{
            onTrainBegin: ()=>{
                console.log('start')
            },
            onTrainEnd: ()=>{
                console.log('done')
            },
            onEpochEnd: async (epoch, logs)=>{
                if( epoch % 100000 === 0 )
                    console.log(epoch, logs.loss)
            }
        }
})
```

this part of the code doesn't work because it should fire after the fit function
```
xs.dispose();
ys.dispose();
```

if you have the opportunity to run the code, you will see everything for yourself

**Comment by mightyplow** [comment]
Oh sorry, my fault. I didn't recognize the amount of epochs. Then you're right and it looks like some internal problem. I'll try it out and see what happens.

**Comment by borodadada** [comment]
Thank you, I'll wait for the result, this problem is bothering me a lot

**Comment by yiv1** [comment]
same problem in Bun

```
2025-06-26 03:02:12.840574: W tensorflow/core/framework/op_kernel.cc:1840] OP_REQUIRES failed at cwise_ops_common.h:124 : RESOURCE_EXHAUSTED: OOM when allocating tensor with shape[32,16] and type float on /job:localhost/replica:0/task:0/device:CPU:0 by allocator mklcpu
============================================================
Bun v1.2.17 (282dda62) Windows x64
Windows v.win10_fe
CPU: sse42 avx avx2
Args: "C:\Users\****\.bun\bin\bun.exe" "src/index.ts"
Features: Bun.stderr Bun.stdout(3) jsc transpiler_cache(9) tsconfig(2) process_dlopen
Builtins: "bun:main" "bun:sqlite" "node:assert" "node:buffer" "node:child_process" "node:crypto" "node:events" "node:fs" "node:os" "node:path" 
"node:stream" "node:url" "node:util"
Elapsed: 52508590ms | User: 54715984ms | Sys: 380281ms
RSS: 4.77GB | Peak: 4.77GB | Commit: 8.95GB | Faults: 80753124

panic(main thread): attempt to unwrap error: OutOfMemory
oh no: Bun has crashed. This indicates a bug in Bun, not your code.
```

There were indeed errors related to memory leaks, not all tensors were cleared.
I fixed them, it got better, but after ~5 training epochs, the crash continues.

I see through `Process Explorer` that `Working set` is stabilized and does not grow infinitely. But `Private bytes` is slowly growing. As far as I could find out, `Private bytes` shows the size of memory allocated in the swap file. And I have no idea how to fix this and how to debug.

**Comment by borodadada** [comment]
I tried to write tests on vitest, using AI, several hours of trial and error and AI wrote the answer that the problem is in the core of the program

**Comment by avmohansai** [comment]
Hello @borodadada ,
Thank you for bringing this to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions ](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute)and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.

**Comment by nileshau-afk** [comment]
Hello @borodadada,

Thank you for opening this issue. Since this issue has been open for a long time, we are closing this issue for now.
If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding.
